design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,148 @@
+It will be available as a domain-resources-requests entry.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 0 and 32.
+- Should be accepted only as `requests` (not as `limits`).
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+#### VMI Spec example
+```yaml
+
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-request-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+    resources:
+      requests:
+        interface: 5
+      rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the resources-requests-interface and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
+
+#### Network Interface Placeholder
+The interface placeholder is a temporary dummy definition of an interface in the domain spec
+of libvirt. It is used to trigger libvirt to auto-generate dependent devices (e.g. PCIe controllers).
+
+After the domain configuration is constructed, these placeholders are removed.
+
+This is an example of such a placeholder:
+```xml
+<interface type='ethernet'>
+  <target dev='placeholder-1' managed='no'/>
+  <model type='virtio-non-transitional'/>
+  <mtu size='1500'/>
+  <rom enabled='no'/>
+</interface>
+```
+
+## Implementation Phases
+N/A
+
+## Appendix A: Alternatives
+There have been several other implementation alternatives to answer the goals.
+This section lists the alternatives to allow an open discussion and record other proposals.
+
+- Allow users to define explicitly domain controllers.
+
+  This option includes the full exposure of the controllers devices details or a partial
+  abstraction of it.
+
+- While using the same suggested API towards the user, construct the required controllers
+  directly (and not through network interface placeholders).
+
+- Use a PCI bridge.
+
+  TBD
```

Comment:
Not sure what this mean, can you please elaborate?

Reply:
tried to express setting 32 pcie controller on each VM everytime would also be an option.

Wouldn't it ?

Reply:
I guess - I actually meant that "as an alternative" to what you're proposing, we could **always** request the maximum number of controllers libvirt allows.

Which would mean each an every VM can be hotplugged with plenty of interfaces.

Reply:
ack
I will add it as part of the next round of changes, thanks!

Reply:
Note that each PCI controller, even if not populated, uses some resources and slows down the guest OS boot. So there's some cost associated with adding unnecessary controllers. Would it still make sense in the context of KubeVirt if it makes the implementation simpler? No idea. The first thing to do would be to measure this overhead...

Reply:
Yes, we're aware it has **a** cost; not sure what it is, since we've never measured it.

I don't think we need to know how much it costs to list this as an alternative though.

Reply:
Done